src/vim-loader/bfast.ts

```diff
@@ -38,116 +94,269 @@ class BFastHeader {
+      ranges.push(new Range(array[i], array[i + 2]))
+    }
+
+    const names = await this.requestNames(ranges[0])
+    if (ranges.length !== names.length + 1) {
+      throw new Error('Mismatched ranges and names count')
+    }
 
-      pos += 4
-      const buffer = new Uint8Array(
-        arrayBuffer,
-        begin + byteOffset,
-        end - begin
-      )
-      buffers.push(buffer)
+    // Map ranges and names
+    const map = new Map<string, Range>()
+    for (let i = 0; i < names.length; i++) {
+      map.set(names[i], ranges[i + 1])
     }
 
-    if (buffers.length < 0) {
-      throw new Error('Expected at least one buffer containing the names')
+    return map
+  }
+
+  private async requestNames (range: Range) {
+    const buffer = await this.request(range, 'Names')
+    const names = new TextDecoder('utf-8').decode(buffer)
+    const result = names.slice(0, -1).split('\0')
+    return result
+  }
+
+  private async requestHeader () {
+    const buffer = await this.request(new Range(0, 32), 'Header')
+    const result = Header.createFromBuffer(buffer)
+    return result
+  }
+
+  private async request (range: Range, label: string): Promise<ArrayBuffer> {
+    // Return from cache if present, then try partial http, fallback to full http
+    const buffer =
+      this.local(range, label) ??
+      (await this.http(range, label)) ??
+      (await this.http(undefined, label))
+
+    if (!buffer) {
+      throw new Error(`Could not load vim at ${this.source}`)
     }
 
-    // break the first one up into names
-    const joinedNames = new TextDecoder('utf-8').decode(buffers[0])
+    if (buffer.byteLength > range.count) {
+      this.source = buffer
+    }
+    return buffer
+  }
 
-    // Removing the trailing '\0' before spliting the names
-    let names = joinedNames.slice(0, -1).split('\0')
-    if (joinedNames.length === 0) names = []
+  private async http (range: Range, label: string) {
+    if (typeof this.source !== 'string') return
+    const xhr = new XMLHttpRequest()
+    xhr.open('GET', this.source)
+    xhr.responseType = 'arraybuffer'
+    const str = `WebRequest ${this.name}.${label}`
 
-    // Validate the number of names
-    if (names.length !== buffers.length - 1) {
-      throw new Error(
-        'Expected number of names to be equal to the number of buffers - 1'
-      )
+    if (range) {
+      const start = this.offset + range.start
+      const end = this.offset + range.end
+      xhr.setRequestHeader('Range', `bytes=${start}-${end - 1}`)
+      console.log(`${str} : [${start}, ${end}] of ${this.source}`)
+    } else {
+      console.log(`${str} of ${this.source}`)
     }
 
-    return new BFast(header, names, buffers.slice(1))
+    return new Promise<ArrayBuffer>((resolve, reject) => {
+      xhr.send()
+      xhr.onprogress = (e) =>
+        console.log(`${str} : ${Math.round(e.loaded / 1000000)} MB`)
+      xhr.onload = () => resolve(xhr.response)
+      xhr.onerror = (_) => {
+        resolve(undefined)
```

Comment:
No. Because one, xmlhttp request already prints a message no matter what happens, and I also throw if both the range request and the fallback full request fail.